common.c

```diff
@@ -65,13 +65,106 @@ int g_enable_tracing = 1;
 struct descriptor_info **g_descriptor_list;
 unsigned int g_descriptor_list_size;
 
+/* special strings to be displayed with ANSI colors */
+struct str_colors {
+	char ch;
+	const char *color_str;
+};
+
+static struct str_colors g_str_colors[] = {
+	{ '\n', VAR "\\n" RST },
+	{ '\r', VAR "\\r" RST },
+	{ '\t', VAR "\\t" RST },
+	{ '\0', NULL }
+};
+
+static void
+trace_printf_str(const char *str)
+{
+	char *str_print;
+
+	rtr_malloc_t	real_malloc;
+	rtr_free_t	real_free;
+	rtr_strncpy_t	real_strncpy;
+	rtr_strcat_t	real_strcat;
+	rtr_strlen_t	real_strlen;
+	rtr_fprintf_t	real_fprintf;
+
+	struct str_colors *colors = g_str_colors;
+
+	/* get original function pointers */
+	real_malloc = RETRACE_GET_REAL(malloc);
+	real_free = RETRACE_GET_REAL(free);
+	real_strncpy = RETRACE_GET_REAL(strncpy);
+	real_strcat = RETRACE_GET_REAL(strcat);
+	real_strlen = RETRACE_GET_REAL(strlen);
+	real_fprintf = RETRACE_GET_REAL(fprintf);
+
+	/* set new buffer */
+	str_print = (char *) real_malloc(strlen(str) + 1);
+	strcpy(str_print, str);
+	str_print[strlen(str)] = '\0';
+
+	while (colors) {
+		char *p = str_print;
+		int idx = 0;
+
+		/* check NULL */
+		if (!colors->color_str)
+			break;
+
+		while (*p != '\0') {
+			if (*p == colors->ch) {
+				char *tmp;
+				size_t size = strlen(str_print) + strlen(colors->color_str) + 1;
```

Comment:
We need to be careful to call our replacements directly as we might fall into loops. Call real_strlen instead.